hypershift-operator/controllers/hostedcluster/karpenter.go

```diff
@@ -92,21 +90,52 @@ spec:
 				},
 				AutoRepair: false,
 			},
-			Platform: hyperv1.NodePoolPlatform{
-				Type: hyperv1.AWSPlatform,
-				AWS: &hyperv1.AWSNodePoolPlatform{
-					InstanceType: "m5.large",
-					Subnet: hyperv1.AWSResourceReference{
-						ID: ptr.To("subnet-none"),
-					},
-				},
-			},
-		}
-		return nil
+			Arch: hyperv1.ArchitectureAMD64, // used to find default AMI
+		},
+	}
+
+	releaseProvider, imageMetadataProvider, err := r.ReconcileMetadataProvidersImpl(ctx, r.RegistryOverrides)
+	if err != nil {
+		return err
+	}
+
+	pullSecretBytes, err := hyperutil.GetPullSecretBytes(ctx, r.Client, hcluster)
+	if err != nil {
+		return err
+	}
+
+	releaseImage, err := releaseProvider.Lookup(ctx, nodePool.Spec.Release.Image, pullSecretBytes)
+	if err != nil {
+		return err
+	}
+
+	haproxy := haproxy.HAProxy{
```

Comment:
Not for this PR but we need to follow up to hide the haproxy generation behind the Config logic.

Then the rest of this logic could be encapsulated within a func token() as in the nodepool_controller if we have a common reconciler interface which implements the methods to get the image

Reply:
can we have this logic to be a unit tested function which validates the expected userData is generated with their current contractual expectations, i.e. hyperv1.NodePoolLabel and hypershift.openshift.io/ami?

Reply:
added a unit test